src/audio/AudioManager.ts

```diff
@@ -0,0 +1,115 @@
+import { useState, useEffect, useMemo } from 'react';
+import { Platform } from 'react-native';
+import { RoomEvent, type Room } from 'livekit-client';
+import AudioSession, {
+  getAppleAudioConfigurationForMode,
+  type AppleAudioConfiguration,
+  type AudioTrackState,
+} from './AudioSession';
+import { log } from '..';
+
+export function useIOSAudioManagement(
+  room: Room,
+  preferSpeakerOutput: boolean = true,
+  onConfigureNativeAudio?: (
+    trackState: AudioTrackState,
+    preferSpeakerOutput: boolean
+  ) => AppleAudioConfiguration
+) {
+  const [localTrackCount, setLocalTrackCount] = useState(0);
+  const [remoteTrackCount, setRemoteTrackCount] = useState(0);
+  const trackState = useMemo(
+    () => computeAudioTrackState(localTrackCount, remoteTrackCount),
+    [localTrackCount, remoteTrackCount]
+  );
+
+  useEffect(() => {
+    if (Platform.OS !== 'ios') {
+      return () => {};
+    }
+
+    setLocalTrackCount(getLocalAudioTrackCount(room));
+    setRemoteTrackCount(getRemoteAudioTrackCount(room));
+
+    let onLocalPublished = () => {
+      log.error('onlocalpub');
+      setLocalTrackCount(localTrackCount + 1);
+    };
+    let onLocalUnpublished = () => {
+      log.error('onlocal unpub');
+      if (localTrackCount - 1 < 0) {
+        log.warn(
+          'mismatched local audio track count! attempted to reduce track count below zero.'
+        );
+      }
+      setLocalTrackCount(Math.max(localTrackCount - 1, 0));
+    };
+    let onRemotePublished = () => {
+      log.error('onremotepub');
+      setRemoteTrackCount(remoteTrackCount + 1);
+    };
+    let onRemoteUnpublished = () => {
+      log.error('onremote unpub');
+      if (remoteTrackCount - 1 < 0) {
+        log.warn(
+          'mismatched remote audio track count! attempted to reduce track count below zero.'
+        );
+      }
+      setRemoteTrackCount(Math.max(remoteTrackCount - 1, 0));
+    };
+
+    room
+      .on(RoomEvent.LocalTrackPublished, onLocalPublished)
+      .on(RoomEvent.LocalTrackUnpublished, onLocalUnpublished)
+      .on(RoomEvent.TrackPublished, onRemotePublished)
+      .on(RoomEvent.TrackUnpublished, onRemoteUnpublished);
+
+    return () => {
+      room
+        .off(RoomEvent.LocalTrackPublished, onLocalPublished)
+        .off(RoomEvent.LocalTrackUnpublished, onLocalUnpublished)
+        .off(RoomEvent.TrackPublished, onRemotePublished)
+        .off(RoomEvent.TrackUnpublished, onRemoteUnpublished);
+    };
+  }, [room, localTrackCount, remoteTrackCount]);
```

Comment:
@lukasIO I'm a little unsure about doing an effect in this manner to track the audio tracks; whenever the track count changes, the effect basically has to `off` the old listeners and `on` the new ones just to update with the new track counts, feels inefficient? Is there a nicer way to do this?

Also if there's any comments about the approach itself, I'm open to ideas 😓

Reply:
generally this looks fine. 
> feels inefficient

that's React's conceptual model for you 😬 

Personally, I would probably try to not pack as much code into a single effect. You could e.g. use `React.useCallback`s for the event handlers and declare them outside of the effect.